core/distributed/partition_helpers.cpp

```diff
@@ -0,0 +1,156 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/distributed/partition.hpp>
+#include <ginkgo/core/distributed/partition_helpers.hpp>
+#include <numeric>
+
+
+#include "core/components/fill_array_kernels.hpp"
+#include "core/components/prefix_sum_kernels.hpp"
+#include "core/distributed/partition_helpers_kernels.hpp"
+
+
+namespace gko {
+namespace experimental {
+namespace distributed {
+namespace components {
+namespace {
+
+
+GKO_REGISTER_OPERATION(fill_seq_array, components::fill_seq_array);
+
+
+}  // namespace
+}  // namespace components
+
+
+namespace partition_helpers {
+namespace {
+
+
+GKO_REGISTER_OPERATION(sort_by_range_start,
+                       partition_helpers::sort_by_range_start);
+GKO_REGISTER_OPERATION(check_consecutive_ranges,
+                       partition_helpers::check_consecutive_ranges);
+
+
+}  // namespace
+}  // namespace partition_helpers
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+std::unique_ptr<Partition<LocalIndexType, GlobalIndexType>>
+build_partition_from_local_range(std::shared_ptr<const Executor> exec,
+                                 mpi::communicator comm, span local_range)
+{
+    std::array<GlobalIndexType, 2> range{
+        static_cast<GlobalIndexType>(local_range.begin),
+        static_cast<GlobalIndexType>(local_range.end)};
+
+    // make all range_start_ends available on each rank
+    auto mpi_exec = exec->get_master();
+    array<GlobalIndexType> ranges_start_end(mpi_exec, comm.size() * 2);
+    ranges_start_end.fill(invalid_index<GlobalIndexType>());
+    std::vector<mpi::request> reqs;
+    reqs.push_back(comm.i_all_gather(mpi_exec, &range[0], 1,
+                                     ranges_start_end.get_data(), 1));
+    reqs.push_back(comm.i_all_gather(
+        mpi_exec, &range[1], 1, ranges_start_end.get_data() + comm.size(), 1));
+    mpi::wait_all(reqs);
+    ranges_start_end.set_executor(exec);
+
+    // make_sort_by_range_start
+    array<comm_index_type> part_ids(exec, comm.size());
+    exec->run(components::make_fill_seq_array(part_ids.get_data(),
+                                              part_ids.get_num_elems()));
+    exec->run(partition_helpers::make_sort_by_range_start(ranges_start_end,
+                                                          part_ids));
+
+    // check for consistency
+    bool consecutive_ranges = false;
+    exec->run(partition_helpers::make_check_consecutive_ranges(
+        ranges_start_end, &consecutive_ranges));
+    if (!consecutive_ranges) {
+        throw Error(__FILE__, __LINE__, "The partition contains gaps.");
+    }
+
+    // remove duplicates
```

Comment:
To me, this description of what happens below could be improved, e.g. 
```suggestion
    // join (now consecutive) starts and ends into combined array
```